File: frontend/src/worker/routes/shortcut.ts
import { Hono } from 'hono'
import { describeShortcutPayloadShape } from '../adapters/shortcutPayload'
import { requireShortcutToken } from '../middleware/auth'
import { shortcutBatchSchema, shortcutTransactionSchema } from '../schemas'
import { TransactionService } from '../services/transactionService'
import type { AppEnvironment, IngestionResult } from '../types'
import { parseBody } from './helpers'

function response(result: IngestionResult) {
  return {
    success: true,
    result: result.result,
    transaction_id: result.transaction.id,
    category: result.transaction.category_name,
    review_status: result.transaction.review_status,
    duplicate: result.duplicate,
  }
}

function logValidationShape(body: unknown) {
  console.warn('Shortcut payload validation rejected', JSON.stringify(describeShortcutPayloadShape(body)))
}

export const shortcutRoutes = new Hono<AppEnvironment>()
shortcutRoutes.use('/shortcut/*', requireShortcutToken)

shortcutRoutes.post('/shortcut/test', (c) => c.json({ success: true }))

shortcutRoutes.post('/shortcut/transactions', async (c) => {
  const payload = await parseBody(c, shortcutTransactionSchema, logValidationShape)
  return c.json(response(await new TransactionService(c.env.DB).ingestShortcut(payload)))
})

shortcutRoutes.post('/shortcut/transactions/batch', async (c) => {
  const payload = await parseBody(c, shortcutBatchSchema)
  const results = []
  for (const item of payload.transactions) results.push(response(await new TransactionService(c.env.DB).ingestShortcut(item)))
  return c.json({
    results,
    created: results.filter((item) => item.result.startsWith('created')).length,
    duplicates: results.filter((item) => item.duplicate).length,
    failed: 0,
  })
})

shortcutRoutes.post('/shortcut/simulate', async (c) => {
  const payload = await parseBody(c, shortcutTransactionSchema, logValidationShape)
  return c.json(response(await new TransactionService(c.env.DB).ingestShortcut(payload, 'simulator', 'shortcut-simulator')))
})

File: backend/app/api/shortcut.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Request
from sqlalchemy.orm import Session

from app.auth.dependencies import require_shortcut_token
from app.database.session import get_db
from app.models.enums import TransactionSource
from app.schemas.transaction import (
    BatchIngestionResponse,
    IngestionResponse,
    ShortcutBatchRequest,
    ShortcutTransactionRequest,
)
from app.services.ingestion import ingest_shortcut
from app.api.presenters import ingestion_response


router = APIRouter(
    prefix="/api/v1/shortcut",
    tags=["shortcut"],
    dependencies=[Depends(require_shortcut_token)],
)


@router.post("/test")
def test_shortcut() -> dict[str, bool]:
    return {"success": True}


@router.post("/transactions", response_model=IngestionResponse)
def create_shortcut_transaction(
    payload: ShortcutTransactionRequest,
    request: Request,
    database: Session = Depends(get_db),
) -> IngestionResponse:
    identity = request.headers.get("Tailscale-User-Login")
    return ingestion_response(
        ingest_shortcut(database, payload, request_identity=identity)
    )


@router.post("/transactions/batch", response_model=BatchIngestionResponse)
def create_shortcut_transactions(
    payload: ShortcutBatchRequest,
    request: Request,
    database: Session = Depends(get_db),
) -> BatchIngestionResponse:
    identity = request.headers.get("Tailscale-User-Login")
    results = [
        ingestion_response(
            ingest_shortcut(database, item, request_identity=identity)
        )
        for item in payload.transactions
    ]
    return BatchIngestionResponse(
        results=results,
        created=sum(item.result.startswith("created") for item in results),
        duplicates=sum(item.duplicate for item in results),
        failed=0,
    )


@router.post("/simulate", response_model=IngestionResponse)
def simulate_transaction(
    payload: ShortcutTransactionRequest,
    database: Session = Depends(get_db),
) -> IngestionResponse:
    return ingestion_response(
        ingest_shortcut(
            database,
            payload,
            source=TransactionSource.SIMULATOR,
            request_identity="local-simulator",
        )
    )
